File: book8-cracking-the-coding-interview/preface-vi-big-o/Example6ReverseArray.c
#include <stdio.h>
#include <stdlib.h>

/*
 * Reverses the array in-place.
 * Time: O(n)
 * Memory: O(1)
 */
void reverse(int *arr, size_t n)
{
    int *p1 = arr;
    int *p2 = arr + n - 1;
    while (p1 < p2)
    {
        int tmp = *p2;
        *p2 = *p1;
        *p1 = tmp;
        p1++;
        p2--;
    }
}

void print_array(const char *prefix, int *arr, int n)
{
    printf("%s", prefix);
    int *p = arr;
    while (--n >= 0)
    {
        printf("%d ", *p++);
    }
    printf("\n");
}

int main(int argc, char *argv[])
{
    int *arr1 = calloc(10, sizeof(int));
    int *p = arr1;
    int i = 0;
    while (p < (arr1 + 10))
    {
        *p++ = i++;
    }
    print_array("Arr1: ", arr1, 10);
    reverse(arr1, 10);
    print_array("Arr1Reversed: ", arr1, 10);
    free(arr1);
    return EXIT_SUCCESS;
}

File: book8-cracking-the-coding-interview/preface-vi-big-o/Example6ReverseArray.java
import java.util.Random;

public class Example6ReverseArray {

    /**
     * Reverses an array.
     * Time: O(n)
     * Memory: O(1) in-place
     * 
     * @param <T>   array element type
     * @param array array to reverse
     */
    <T> void reverse(T[] array) {
        for (int i = 0; i < array.length / 2; i++) {
            int other = array.length - i - 1;
            T temp = array[i];
            array[i] = array[other];
            array[other] = temp;
        }
    }

    void run() {
        Integer[] arr1 = new Integer[] { 1, 2, 3, 4, 5 };
        printArray("Array1: ", arr1);
        reverse(arr1);
        printArray("Array1 Reversed: ", arr1);
        Integer[] arr2 = new Integer[20];
        for (int i = 0; i < arr2.length; i++) {
            arr2[i] = new Random().nextInt(100);
        }
        printArray("Array2: ", arr2);
        reverse(arr2);
        printArray("Array2 Reversed: ", arr2);
    }

    public static void main(String[] args) {
        new Example6ReverseArray().run();
    }

    <T> void printArray(String prefix, T[] arr) {
        System.out.print(prefix);
        for (T i : arr) {
            System.out.print(i);
            System.out.print(' ');
        }
        System.out.println();
    }
}